amaranth/vendor/_lattice_nexus.py

```diff
@@ -0,0 +1,884 @@
+    _radiant_file_templates = {
+        **TemplatedPlatform.build_script_templates,
+        "build_{{name}}.sh": r"""
+            # {{autogenerated}}
+            set -e{{verbose("x")}}
+            if [ -z "$BASH" ] ; then exec /bin/bash "$0" "$@"; fi
+            {{emit_commands("sh")}}
+        """,
+        "{{name}}.v": r"""
+            /* {{autogenerated}} */
+            {{emit_verilog()}}
+        """,
+        "{{name}}.debug.v": r"""
+            /* {{autogenerated}} */
+            {{emit_debug_verilog()}}
+        """,
+        "{{name}}.tcl": r"""
+            prj_create -name {{name}} -impl impl \
+                -dev {{platform.device}}-{{platform.speed}}{{platform.package}}{{platform.grade}} \
+                -synthesis synplify
+            {% for file in platform.iter_files(".v", ".sv", ".vhd", ".vhdl") -%}
+                prj_add_source {{file|tcl_quote}}
+            {% endfor %}
+            prj_add_source {{name}}.v
+            prj_add_source {{name}}.sdc
+            prj_add_source {{name}}.pdc
+            prj_set_impl_opt top \"{{name}}\"
+            {{get_override("script_project")|default("# (script_project placeholder)")}}
+            prj_save
+            prj_run Synthesis -impl impl -forceOne
+            prj_run Map -impl impl
+            prj_run PAR -impl impl
+            prj_run Export -impl impl -task Bitgen
+            {{get_override("script_after_export")|default("# (script_after_export placeholder)")}}
+        """,
+        # Pre-synthesis SDC constraints
+        "{{name}}.sdc": r"""
+            {% for net_signal, port_signal, frequency in platform.iter_clock_constraints() -%}
+                {% if port_signal is not none -%}
+                    create_clock -name {{port_signal.name|tcl_quote}} -period {{1000000000/frequency}} [get_ports {{port_signal.name}}]
+                {% else -%}
+                    create_clock -name {{net_signal.name|tcl_quote}} -period {{1000000000/frequency}} [get_nets {{net_signal|hierarchy("/")}}]
+                {% endif %}
+            {% endfor %}
+            {{get_override("add_constraints")|default("# (add_constraints placeholder)")}}
+        """,
+        # Physical PDC contraints
+        "{{name}}.pdc": r"""
+            {% for port_name, pin_name, attrs in platform.iter_port_constraints_bits() -%}
+                ldc_set_location -site "{{pin_name}}" [get_ports {{port_name|tcl_quote}}]
+                {% if attrs -%}
+                ldc_set_port -iobuf { {%- for key, value in attrs.items() %} {{key}}={{value}}{% endfor %} } [get_ports {{port_name|tcl_quote}}]
+                {% endif %}
+            {% endfor %}
+            {{get_override("add_preferences")|default("# (add_preferences placeholder)")}}
+        """,
+    }
+    _radiant_command_templates = [
+        # These don't have any usable command-line option overrides.
+        r"""
+        {{invoke_tool("radiantc")}}
+            {{name}}.tcl
+        """,
+        ## TODO: FIXME
```

Comment:
What's the status of this `FIXME`? It looks like you're doing bitgen as a part of the main Tcl script?

Reply:
FIXME was intended to do what lattice ecp5 (Diamond) does (namely, copy bitstream to build/top.bit and program via lattice tools. However, haven't tested this with Radiant (just using ecpprog as of now). Removed in [8ccc083](https://github.com/amaranth-lang/amaranth/pull/759/commits/8ccc083c024f48991b2218ba57ba6bb91e33bc49). Yes, bitgen is happening in main Tcl script.